src/witness/yamlWitness.ml

```diff
@@ -174,43 +218,107 @@ struct
             | None ->
               acc
           end
-        | _ -> (* avoid FunctionEntry/Function because their locations are not inside the function where assert could be inserted *)
+        end else begin
+          (* avoid FunctionEntry/Function because their locations are not inside the function where assert could be inserted *)
           acc
+        end
       ) nh []
     in
 
-    (* TODO: deduplicate *)
+
+    (* Generate precondition invariants.
+       We do this in three steps:
+       1. Collect contexts for each function
+       2. For each function context, find "matching"/"weaker" contexts that may satisfy its invariant
+       3. Generate precondition invariants. The postcondition is a disjunction over the invariants for matching states. *)
+
+    (* 1. Collect contexts for each function *)
```

Comment:
Slightly annoying that #391 isn't merged yet, because there's an `IterSysVars` query that directly returns this. Maybe add a TODO here to use that once that's merged.